notion-server/src/ethereum.rs

```diff
@@ -1,4 +1,77 @@
+use base64;
 use std::collections::HashMap;
+use std::convert::TryFrom;
+use snafu::{Snafu, ResultExt, Backtrace};
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    Decode {
+        backtrace: Backtrace,
+        source: base64::DecodeError,
+    },
+    OutOfBounds {
+        message: String,
+    }
+}
+
+pub mod interface_args {
+    #[derive(Debug, Default)]
+    pub struct GeneralSimulationState {
+        pub num_execution_environments: u32,
+        pub num_shard_chains: u32,
+    }
+
+    #[derive(Debug, Default)]
+    pub struct ExecutionEnvironment {
+        pub base64_encoded_wasm_code: String,
+    }
+
+    impl From<&super::ExecutionEnvironment> for ExecutionEnvironment {
+        fn from(ee: &super::ExecutionEnvironment) -> Self {
+            let base64_encoded_wasm_code = base64::encode(&ee.wasm_code);
+            Self {
+                base64_encoded_wasm_code,
+            }
+        }
+    }
+
+    #[derive(Debug, Default)]
+    pub struct CreateShardChain {}
+
+    #[derive(Debug, Default, Eq, PartialEq)]
+    pub struct ShardBlock {
+        pub transactions: Vec<ShardTransaction>,
+    }
+    impl From<&super::ShardBlock> for ShardBlock {
+        fn from(sb: &super::ShardBlock) -> Self {
+            let transactions: Vec<ShardTransaction> = sb.transactions.iter().map(|st| -> ShardTransaction {
+                ShardTransaction::from(st)
+            }).collect();
+            ShardBlock {
+                transactions,
+            }
+        }
+    }
+
+    #[derive(Debug, Default, Eq, PartialEq)]
+    pub struct ShardTransaction {
+        pub base64_encoded_data: String,
+        pub ee_index: u32,
```

Comment:
Are we going with typed execution environment indexes (i.e. `struct EeIndex(i32)`) or no?

Reply:
My intent here was to keep what I considered "internal" structs / data structures completely out of the public-facing interface.  So in this case: while `EeIndex` seemed like a great way to make our code more clear and easier to read for writing internal logic, IMO the interface will be more accessible and will be easier to use if all the external-facing types are (for the most part) either 1) common types users are used to seeing (eg. u32, String, possible Vec<u8>, etc), or 2) custom types useful for defining the interface or bundling function arguments (eg. `ShardTransaction` and the other structs in `mod interface_args`.

Thanks for the callout, definitely interested in discussing with you more, let's chat about this next time we check in.

Reply:
I think whatever arguments apply for using `EeIndex` internally also apply externally.

These kind of types are called `newtypes` and are fairly useful to organize code, and prevent accidentally passing a shard index into a method expecting an ee index, for example.